# ALFRED_method Overview

This repository contains instructions for analysis of "Systematic discovery of germline cancer predisposition genes through the identification of somatic second hits" by Park et al (https://www.nature.com/articles/s41467-018-04900-7). In general code for LOH calling from cancer samples and randomization for measuring the excess of germline variants (rare damaging germline variants) in LOH samples (or cancer samples) compared to samples without LOH events (or non-cancer samples).

Abstract
The genetic causes of cancer include both somatic mutations and inherited germline variants. Large-scale tumor sequencing has revolutionized the identification of somatic driver alterations but has had limited impact on the identification of cancer predisposition genes (CPGs). Here we present a statistical method, ALFRED, that tests Knudson’s two-hit hypothesis to systematically identify CPGs from cancer genome data. Applied to ~10,000 tumor exomes the approach identifies known and putative CPGs – including the chromatin modifier NSD1 – that contribute to cancer through a combination of rare germline variants and somatic loss-of-heterozygosity (LOH). Rare germline variants in these genes contribute substantially to cancer risk, including to ~14% of ovarian carcinomas, ~7% of breast tumors, ~4% of uterine corpus endometrial carcinomas, and to a median of 2% of tumors across 17 cancer types.
